File: 03 哈希表/番外-水仙花数.py
# 跟快乐数不同，水仙花数的判断不需要循环，可以一步完成，跟哈希表也没关系，只是写快乐数的时候联想到了复习一下这个非常初级的问题

def is_narcissistic(num: int) -> bool:
    n = len(str(num))   # 位数
    new_sum = 0
    for i in str(num):
        new_sum += int(i) ** n

    if new_sum == num:
        return True
    else:
        return False

print(is_narcissistic(154))

File: 03 哈希表/番外-水仙花数.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>

// 跟python不同，c原生支持的函数不够丰富，繁琐一些，c不支持直接把数转为字符串，所以用到循环较多
bool is_narcissistic(int num)
{
    int n = 0;
    int new_sum = 0;

    // 计算位数n
    int temp = n;
    while(temp != 0)
    {
        temp /= 10;
        ++n;
    }

    temp = num;
    while(temp != 0)
    {
        int digit = temp % 10;  // 取余数，最后一位
        new_sum += (int)pow(digit, n);
        temp /= 10;
    }
    if(new_sum == num)
    {
        return true;
    }
    else{
        return false;
    }
}

void main()
{
    printf("%d\n", is_narcissistic(153));
}